### PR TITLE
[Shadow CSS] Host context fix

### PR DESCRIPTION
**Pull request by alx1417**
## PR Checklist
Please check if your PR fulfills the following requirements:

- [x] The commit message follows our guidelines: https://github.com/angular/angular/blob/master/CONTRIBUTING.md#commit
- [x] Tests for the changes have been added (for bug fixes / features)
- [x] Docs have been added / updated (for bug fixes / features)

## PR Type
What kind of change does this PR introduce?

```
[x] Bugfix
[ ] Feature
[ ] Code style update (formatting, local variables)
[ ] Refactoring (no functional changes, no api changes)
[ ] Build related changes
[ ] CI related changes
[ ] Documentation content changes
[ ] angular.io application / infrastructure changes
[ ] Other... Please describe:
```

## What is the current behavior?

When  you use more than one :host-context selector in the same css rule, the compiled css generate a invalid css code "-shadowcsscontext".

Example: 
````scss
// Preprocessed code
:host-context(.one):host-context(.two) {}
````
````css
/* Generated code*/
.one-shadowcsscontext(.two)[nghost],
.one   -shadowcsscontext(.two)[nghost] {}
````

Issue Number: https://github.com/angular/angular/issues/19199

## What is the new behavior?

You can use two or more :host-context selector in the same css rule and generate all css combinations.

Example:

````scss
// Preprocessed code
:host-context(.one) :host-context(.two) {}
````
````css
/* Generated code*/
.one[nghost] .two[ngcontent],
.one .two [a-host], 
.two[nghost] .one[ngcontent], 
.two .one [nghost], 
.one.two[nghost],
.one.two [nghost] {}
````

## Does this PR introduce a breaking change?
```
[ ] Yes
[x] No
```

### CODE REVIEW

**Comment by alx1417** [comment]
I signed it!

**Comment by alx1417** [comment]
I signed it!

**Comment by alx1417** [comment]
I signed it!

**Comment by vicb** [comment]
What does `:host-context(.one):host-context(.two)` means ? is it equivalent to `:host-context(.one .two)` ?

**Comment by alx1417** [comment]
Hi vicb,

The code that you propouse is similar but not equivalent.

When use `:host-context(.one .two)` only have one combination of class (.one>.two).

When use `:host-context(.one):host-context(.two)` you have the two possible orders(.one>.two or .two >.one).

Also, in some cases is not possible use  `:host-context(.one .two)` : 
- You not know the order of the clases. `.one .two {} OR .two .one {}`
- The structure of the scss does not allow it, for example using the host context in multiples @mixins 

````scss
// mixins.scss
@mixin device-type-mobile {
  :host-context(.mobile) & {  @content;  }
}
@mixin device-os-ios {
  :host-context(.ios) & { @content;  }
}
// usage.scss
.wrapper {
  background: green;
  @include device-type-mobile {
    background: red;
    @include device-os-ios {
      background: blue;
    }
  }
}
// generated.css (Forced break lines and commented)

 // No mixin applied.
.wrapper[_ngcontent-c5] {
  background: green;
}

// One mixin applied
.mobile[_nghost-c5]   .wrapper[_ngcontent-c5], 
.mobile   [_nghost-c5]   .wrapper[_ngcontent-c5] {
  background: red;
}

// Two mixin applied
// Order .ios > .mobile
.ios[_nghost-c5]   .mobile[_ngcontent-c5]   .wrapper[_ngcontent-c5],
.ios   .mobile   [_nghost-c5]   .wrapper[_ngcontent-c5],
// Order .mobile > .ios
.mobile[_nghost-c5]   .ios[_ngcontent-c5]   .wrapper[_ngcontent-c5], 
.mobile   .ios   [_nghost-c5]   .wrapper[_ngcontent-c5],
// Same level .mobile = .ios
.ios.mobile[_nghost-c5]   .wrapper[_ngcontent-c5],
.ios.mobile   [_nghost-c5]   .wrapper[_ngcontent-c5] {
  background: blue;
}
````

Thank you for review the code.

**Comment by vicb** [comment]
> When use :host-context(.one):host-context(.two) you have the two possible orders(.one>.two or .two >.one).

Do you have a link to a spec that would explain this behavior ?

**Comment by alx1417** [comment]
> Do you have a link to a spec that would explain this behavior ?

In this [link](https://angular.io/guide/component-styles#host-context) the explain of the :host-context say:

> The :host-context() selector looks for a CSS class in any ancestor of the component host element, up to the document root.

Not specify the behaviour when use two host-context, but if we apply the previous description to two host context, we have two selectors that "looks for a CSS class in any ancestor".

````scss
:host-context(.one) {} // .one [host] {}
:host-context(.two) {} // .two [host] {}
:host-context(.one .two) {} // .one .two [host] {}
:host-context(.two .one ) {} // .two .one [host] {}

:host-context(.one):host-context(.two) {} // .one .two [host], .two .one [host] {}
// .one should be ancestor and .two also
// NOTE: Now, before the fix, this return an invalid css (.one -shadowcsscontext(.two)  [host])

:host-context(.one) {
  color: red;
  :host-context(.two) {
    color: blue;
  }
}
/* Is not necessary that stay at the same level, for example
.one [host] {color: red},
.one .two [host], .two .one[host] {color: blue}
*/
````

I generate the two orders because the two host-context should by ancestors.

File: packages/compiler/test/shadow_css_spec.ts [hunk 1]
```diff
@@ -195,6 +195,12 @@ export function main() {
         expect(s(':host-context([a=b]) {}', 'contenta', 'a-host'))
             .toEqual('[a=b][a-host], [a="b"] [a-host] {}');
       });
+
+      it('should handle multiple host-context', () => {
+        expect(s(':host-context(.one) :host-context(.two) {}', 'contenta', 'a-host'))
+            .toEqual(
+                '.one[a-host] .two[contenta], .one .two [a-host], .two[a-host] .one[contenta], .two .one [a-host], .one.two[a-host], .one.two [a-host] {}');
+      });
```
Comment (petebacondarwin) [inline packages/compiler/test/shadow_css_spec.ts:203]: Just rebasing and tidying up this PR. I think this test is not actually correct. 

First, I don't think that there should be a space between the `:host-context()` selectors in the string passed to `s()`. Having `:host-context(...) :host-context(...)` doesn't make sense to me. I am not even sure how we should interpret this.

Second, I think that `contenta` should not appear in the final selection at all.

This is what I would expect:

```suggestion
expect(s(':host-context(.one):host-context(.two) {}', 'contenta', 'a-host')
            .replace(/ \{\}$/, '')
            .split(/\,\s+/)
            .sort())
    .toEqual([
      '.one .two[a-host]',   // `one` is an ancestor and `two` is on the host
      '.one .two [a-host]',  // `one` and `two` are both ancestors (in that order)
      '.one.two [a-host]',   // `one` and `two` are both on the same ancestor
      '.one.two[a-host]',    // `one` and `two` both on the host
      '.two .one [a-host]',  // `two` and `one` are both ancestors (in that order)
      '.two .one[a-host]',   // `two` is an ancestor and `one` is on the host
    ]);
```

**Review by petebacondarwin — CHANGES_REQUESTED** [review]
First of all, thank you @alx1417 for this PR. I am sorry it has taken so long to review it.

I don't believe the test nor the implementation are correct here.

Can you consider my suggested test? If you agree the test is correct, then the implementation needs to change since it fails this test.

Further, if you are going to update the implementation, then I think the changes should be in the `_colonHostContextPartReplacer()` method, rather than the `_convertColonRule()` which is shared with handling `:host()` selectors. And we should also look at the array manipulation being used as I think this could be a bit more performant.

**Comment by petebacondarwin** [comment]
Closing in favour of #40494.

**Pull request closed without merging**